File: frontend/src/types/project.ts
/**
 * Frontend mirror of the backend Project DTO (docs/product-spec.md §6).
 *
 * The frontend never computes dates, working days, successors or effort — those
 * arrive already resolved from the backend. Keeping this type identical to the
 * DTO means the prototype's fixtures can later be swapped for real API
 * responses without touching any component.
 */

export type CreatedSource = 'import' | 'agent'

/** ISO date, `YYYY-MM-DD`. Date-only, no timezone. */
export type IsoDate = string

export interface Task {
  id: string
  name: string
  description: string
  assignee: string | null
  duration_workdays: number
  planned_effort_hours: number
  predecessor_ids: string[]
  /** Derived server-side; present on the DTO, never recomputed here. */
  successor_ids: string[]
  start_not_before: IsoDate | null
  start_date: IsoDate
  end_date: IsoDate
  display_order: number
  created_source: CreatedSource
}

export interface Project {
  id: string
  name: string
  project_start_date: IsoDate
  revision: number
  tasks: Task[]
  created_at: string
  updated_at: string
}

/* -----------------------------------------------------------------------------
 * AI change semantics
 * -------------------------------------------------------------------------- */

/**
 * How a task was touched by the last AI command.
 *
 * Precedence when a task qualifies for more than one (technical-blueprint §6):
 *   created > direct > derived
 */
export type ChangeType = 'direct' | 'created' | 'derived'

/** A single field-level difference shown in the change summary. */
export interface FieldDelta {
  label: string
  before: string
  after: string
}

export interface ModifiedTaskChange {
  taskId: string
  name: string
  deltas: FieldDelta[]
}

export interface CreatedTaskChange {
  taskId: string
  name: string
  assignee: string | null
  durationWorkdays: number
  plannedEffortHours: number
}

export interface DerivedTaskChange {
  taskId: string
  name: string
  beforeStart: IsoDate
  afterStart: IsoDate
  beforeEnd: IsoDate
  afterEnd: IsoDate
  /** Shift of the end date, in working days. Positive = later. */
  workdayShift: number
}

/** A dependency edge that the change set added or removed. */
export interface LinkChange {
  sourceId: string
  targetId: string
  kind: 'added' | 'removed'
}

export interface ReleaseImpact {
  taskId: string
  taskName: string
  beforeDate: IsoDate
  afterDate: IsoDate
  /** Working days between the two release dates. Positive = slipped. */
  workdayShift: number
}

/**
 * The full diff between two project revisions. Produced by `computeChangeSet`
 * from before/after DTOs, so it stays correct if the backend ever returns a
 * different result than the canonical demo.
 */
export interface ChangeSet {
  fromRevision: number
  toRevision: number
  /** Change type per task id — the index the Gantt reads for highlighting. */
  byTask: Record<string, ChangeType>
  modified: ModifiedTaskChange[]
  created: CreatedTaskChange[]
  derived: DerivedTaskChange[]
  links: LinkChange[]
  /** Previous bar position per task, for the "was here" ghost outline. */
  previousSchedule: Record<string, { start: IsoDate; end: IsoDate }>
  release: ReleaseImpact | null
  taskCountBefore: number
  taskCountAfter: number
}

File: backend/app/schemas/project.py
"""API-facing Project/Task shapes.

This is the one seam where `successor_ids` gets computed — it is
deliberately *not* part of the domain `Task` (see app/domain/models.py) so
nothing internal can mistake it for authoritative per-task state. Domain
`Project`/`Task` objects are never returned directly from a router.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel

from app.domain.models import CreatedSource, Project


class TaskOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str
    assignee: str | None
    duration_workdays: int
    planned_effort_hours: int
    predecessor_ids: list[uuid.UUID]
    successor_ids: list[uuid.UUID]
    start_not_before: date | None
    start_date: date
    end_date: date
    display_order: int
    created_source: CreatedSource


class ProjectOut(BaseModel):
    id: uuid.UUID
    name: str
    project_start_date: date
    revision: int
    tasks: list[TaskOut]
    created_at: datetime
    updated_at: datetime


class ImportResponse(BaseModel):
    project: ProjectOut
    warnings: list[str] = []


def project_to_out(project: Project) -> ProjectOut:
    successors: dict[uuid.UUID, list[uuid.UUID]] = {task.id: [] for task in project.tasks}
    for task in project.tasks:
        for pred_id in task.predecessor_ids:
            if pred_id in successors:
                successors[pred_id].append(task.id)

    ordered = sorted(project.tasks, key=lambda t: t.display_order)
    tasks_out = [
        TaskOut(
            id=task.id,
            name=task.name,
            description=task.description,
            assignee=task.assignee,
            duration_workdays=task.duration_workdays,
            planned_effort_hours=task.planned_effort_hours,
            predecessor_ids=task.predecessor_ids,
            successor_ids=successors[task.id],
            start_not_before=task.start_not_before,
            start_date=task.start_date,
            end_date=task.end_date,
            display_order=task.display_order,
            created_source=task.created_source,
        )
        for task in ordered
    ]

    return ProjectOut(
        id=project.id,
        name=project.name,
        project_start_date=project.project_start_date,
        revision=project.revision,
        tasks=tasks_out,
        created_at=project.created_at,
        updated_at=project.updated_at,
    )
